feat: enable xpu support for meta-reference stack

dvrogozh opened the pull request:
This commit adds support for XPU and CPU devices into meta-reference stack for text models. On creation stack automatically identifies which device to use checking available accelerate capabilities in the following order: CUDA, then XPU, finally CPU. This behaviour can be overwritten with the `DEVICE` environment variable. In this case explicitly specified device will be used.

Tested with:
```
torchrun pytest llama_stack/providers/tests/inference/test_text_inference.py -k meta_reference
```

Results:
* Tested on: system with single CUDA device, system with single XPU device and on pure CPU system
* Results: all test pass except `test_completion_logprobs`
* `test_completion_logprobs` fails in the same way as on a baseline, i.e. unrelated with this change: `AssertionError: Unexpected top_k=3`

Requires: https://github.com/meta-llama/llama-models/pull/233

dvrogozh commented:
@ashwinb : can you, please, give some insights on the existing tests? At the moment there is a lack of documentation regarding tests in llama-stack. Can you suggest if there are existing tests for meta-reference stack which should be extended to work in non-cuda environments?

I thought that https://github.com/meta-llama/llama-stack/blob/main/llama_stack/providers/inline/agents/meta_reference/tests/test_chat_agent.py might provide such a test. However, it fails for me running even in non-modified llama-stack on CUDA system. Questions:
* Is this test up to date and can run and pass?
* Does this test require built and started meta-reference-gpu distribution?

I am getting this when run on-modified llama-stack on CUDA system:
```
pytest -rsf ./llama_stack/providers/inline/agents/meta_reference/tests
...
E           TypeError: MockToolGroupsAPI.list_tools() got an unexpected keyword argument 'toolgroup_id'
...
FAILED llama_stack/providers/inline/agents/meta_reference/tests/test_chat_agent.py::test_chat_agent_create_and_execute_turn - pydantic_core._pydantic_core.ValidationError: 1 validation error for ChatCompletionResponseEvent
FAILED llama_stack/providers/inline/agents/meta_reference/tests/test_chat_agent.py::test_chat_agent_complex_turn - pydantic_core._pydantic_core.ValidationError: 1 validation error for ChatCompletionResponseEvent
FAILED llama_stack/providers/inline/agents/meta_reference/tests/test_chat_agent.py::test_chat_agent_tools[toolgroups1-True-False] - TypeError: MockToolGroupsAPI.list_tools() got an unexpected keyword argument 'toolgroup_id'
FAILED llama_stack/providers/inline/agents/meta_reference/tests/test_chat_agent.py::test_chat_agent_tools[toolgroups2-False-True] - TypeError: MockToolGroupsAPI.list_tools() got an unexpected keyword argument 'toolgroup_id'
FAILED llama_stack/providers/inline/agents/meta_reference/tests/test_chat_agent.py::test_chat_agent_tools[toolgroups3-True-True] - TypeError: MockToolGroupsAPI.list_tools() got an unexpected keyword argument 'toolgroup_id'
===================================== 5 failed, 2 passed, 10 warnings in 1.35s =====================================
```

ashwinb commented:
Will kill that test, it is stale. The test setup that is relevant for you is in `providers/tests/inference`. See https://github.com/meta-llama/llama-stack/blob/main/llama_stack/providers/tests/README.md

On a CUDA system, you can run it like this:
```
pytest -s -v llama_stack/providers/tests/inference/test_text_inference.py \
  -m "meta-reference and llama_8b" \
  --env <...>
```

ashwinb reviewed (approved):
Thank you, this is going to be quite useful.

dvrogozh commented:
@ashwinb : thank you for pointing out the relevant test. I have rebased the PR on the latest main. The test you pointed to works for non-CUDA device without modifications since at the moment I added automated device selection on the model creation.

FYI, I observe the following test failure on the main code without applying this PR:
```
# torchrun `which pytest` llama_stack/providers/tests/inference/test_text_inference.py -k "meta_reference and test_completion_logprobs"
...
llama_stack/providers/tests/inference/test_text_inference.py:145: in test_completion_logprobs
    response = await inference_impl.completion(
llama_stack/providers/utils/telemetry/trace_protocol.py:101: in async_wrapper
    result = await method(self, *args, **kwargs)
llama_stack/distribution/routers/routers.py:194: in completion
    return await provider.completion(**params)
llama_stack/providers/utils/telemetry/trace_protocol.py:101: in async_wrapper
    result = await method(self, *args, **kwargs)
llama_stack/providers/inline/inference/meta_reference/inference.py:149: in completion
    assert logprobs.top_k == 1, f"Unexpected top_k={logprobs.top_k}"
E   AssertionError: Unexpected top_k=3
...
=
FAILED llama_stack/providers/tests/inference/test_text_inference.py::TestInference::test_completion_logprobs[-meta_reference] - AssertionError: Unexpected top_k=3

```
There seems to be a corresponding enhancement request:
* https://github.com/meta-llama/llama-stack/issues/214

ashwinb commented:
@dvrogozh yes the implementation only supports top_k but the test is using top_k=3, will fix
